llvm/lib/Target/RISCV/RISCVISelLowering.cpp

```diff
@@ -13689,6 +13695,221 @@ static SDValue performXORCombine(SDNode *N, SelectionDAG &DAG,
   return combineSelectAndUseCommutative(N, DAG, /*AllOnes*/ false, Subtarget);
 }
 
+struct Step {
+  unsigned Opcode;
+  uint64_t A;
+  uint64_t B;
+  unsigned Shift = 0;
+};
+
+static bool findMulExpansionRecursive(uint64_t MulAmt, bool HasShlAdd,
+                                      SmallVector<Step> &Path) {
```

Comment:
Can we do decision tree pruning here?
```suggestion
                                      SmallVector<Step> &Path, uint32_t BestPathSize) {
      if (Path.size() + 1 >= BestPathSize)
          return false;
```